Data race in service-mirror RemoteClusterServiceWatcher

### What is the issue?

`RemoteClusterServiceWatcher` in `multicluster/service-mirror/cluster_watcher.go` has no synchronization on its shared receiver. Multiple goroutines launched by `Start()` concurrently access and mutate `rcsw` fields without a mutex.


## Affected Code

**File:** `multicluster/service-mirror/cluster_watcher.go`
**Component:** service-mirror (multicluster)
**Severity:** Low-medium (event queue serializes most work, but direct field writes race)

## The Race

`Start()` launches two goroutines that access `rcsw`:

### Goroutine 1: `processEvents` (line 1459)
```go
go rcsw.processEvents(ctx)
```
Processes events from `eventsQueue`, calls handlers like `createOrUpdateService`, `handleCreateOrUpdateEndpoints`, `repairEndpoints` — all reading/writing `rcsw` fields.

### Goroutine 2: repair ticker (line 1471)
```go
go func() {
    ticker := time.NewTicker(rcsw.repairPeriod)
    for {
        select {
        case <-ticker.C:
            ev := RepairEndpoints{}
            rcsw.eventsQueue.Add(&ev)
        case alive := <-rcsw.liveness:
            rcsw.gatewayAlive = alive  // ← DIRECT FIELD WRITE, NO LOCK
            ev := RepairEndpoints{}
            rcsw.eventsQueue.Add(&ev)
        case <-rcsw.stopper:
            return
        }
    }
}()
```

**Line 1479:** `rcsw.gatewayAlive = alive` is a direct field write from the repair goroutine while `processEvents` may be reading `rcsw.gatewayAlive` in `repairEndpoints()` or other handlers.

### Affected methods (both files)

All methods on `RemoteClusterServiceWatcher` share the unprotected receiver across goroutines. This includes methods in both `cluster_watcher.go` and `cluster_watcher_headless.go`:

- `cluster_watcher.go`: `repairEndpoints`, `createOrUpdateGatewayEndpoints`, `handleCreateOrUpdateEndpoints`, `cleanupOrphanedServices`, `cleanupMirroredResources`
- `cluster_watcher_headless.go`: `createOrUpdateHeadlessEndpoints`, `createHeadlessMirrorEndpoints`, `createEndpointMirrorService`

### Additionally

Informer callbacks registered in `Start()` (lines 1374-1458) run on Kubernetes informer goroutines. They enqueue events to `eventsQueue` (thread-safe), but the Informer goroutines are a third concurrent accessor of the `rcsw` struct.

## Impact

- `rcsw.gatewayAlive` read/write race between repair goroutine and processEvents
- Stale `gatewayAlive` value → endpoints repaired against wrong gateway state → **traffic routed to dead gateways** in multicluster setups
- Under gateway flapping (rapid liveness changes), the race window grows
- Potential panic from nil pointer if handlers assume consistent state during concurrent mutation

## Precedent: #11163 (same pattern, different component)

This is the same race pattern that caused production panics in the Destination endpoint watcher (#11163, reported via #11135). In that case:

- `pp.updateServer()` called `Add` on listeners with no mutex
- `pp.unsubscribe()` set listeners to nil concurrently
- Result: **nil pointer panic in production** under high endpoint churn
- Fix: added mutex protection to all functions that modify listeners

The service-mirror `RemoteClusterServiceWatcher` has the same architectural gap — shared struct accessed from multiple goroutines with zero synchronization. The Destination component was fixed; the service-mirror component was not.

## Detection

Found by [DryRun](https://system-design.app/), confirmed by `go test -race`.

## Verification

- **Confirmed in:** stable-2.12.3 
- **Confirmed in:** latest main (0 mutex/lock calls in cluster_watcher.go as of 2026-03-25)
- **No existing issues** found for this race
- **Different subsystem** from #11163 (which is in Destination/endpoints_watcher)

### How can it be reproduced?

The race is confirmed by `go test -race`. Add this test to `cluster_watcher_race_test.go`:

```go
func TestGatewayAliveRace(t *testing.T) {
	liveness := make(chan bool, 100)
	stopper := make(chan struct{})

	watcher := &RemoteClusterServiceWatcher{
		gatewayAlive: true,
		liveness:     liveness,
		stopper:      stopper,
	}

	// Goroutine 1: reads gatewayAlive (simulates processEvents/repairEndpoints)
	go func() {
		for {
			select {
			case <-stopper:
				return
			default:
				_ = watcher.gatewayAlive
				time.Sleep(time.Millisecond)
			}
		}
	}()

	// Goroutine 2: writes gatewayAlive (simulates repair ticker liveness handler)
	go func() {
		for {
			select {
			case alive := <-liveness:
				watcher.gatewayAlive = alive
			case <-stopper:
				return
			}
		}
	}()

	for i := 0; i < 200; i++ {
		liveness <- (i%2 == 0)
	}

	time.Sleep(50 * time.Millisecond)
	close(stopper)
}
```

### Logs, error output, etc

```
$ go test -race -run TestGatewayAliveRace ./multicluster/service-mirror/
WARNING: DATA RACE
Write at cluster_watcher_race_test.go:41 by goroutine 60
Previous read at cluster_watcher_race_test.go:30 by goroutine 59
--- FAIL: TestGatewayAliveRace (0.05s)
FAIL
```

### output of `linkerd check -o short`

N/A — static analysis, no cluster.

### Environment

N/A — found via static analysis of source code, not runtime testing.

### Possible solution

Add a `sync.Mutex` to `RemoteClusterServiceWatcher`:

```go
type RemoteClusterServiceWatcher struct {
    mu           sync.Mutex  // protects gatewayAlive and other shared fields
    // ... existing fields
}
```

Protect the direct write:
```go
case alive := <-rcsw.liveness:
    rcsw.mu.Lock()
    rcsw.gatewayAlive = alive
    rcsw.mu.Unlock()
    // ...
```

And reads in handlers:
```go
func (rcsw *RemoteClusterServiceWatcher) repairEndpoints(ctx context.Context) error {
    rcsw.mu.Lock()
    alive := rcsw.gatewayAlive
    rcsw.mu.Unlock()
    // use alive...
}
```

Alternatively, use `atomic.Bool` for `gatewayAlive` since it's a single boolean field.

### Additional context

_No response_

### Would you like to work on fixing this bug?

yes

--- a/multicluster/service-mirror/cluster_watcher.go
+++ b/multicluster/service-mirror/cluster_watcher.go
@@ -10,6 +10,7 @@ import (
 	"sort"
 	"strconv"
 	"strings"
+	"sync/atomic"
 	"time"
 
 	"github.com/linkerd/linkerd2/controller/gen/apis/link/v1alpha3"
@@ -60,7 +61,7 @@ type (
 		eventsQueue              workqueue.TypedRateLimitingInterface[any]
 		requeueLimit             int
 		repairPeriod             time.Duration
-		gatewayAlive             bool
+		gatewayAlive             atomic.Bool
 		liveness                 chan bool
 		headlessServicesEnabled  bool
 		namespaceCreationEnabled bool
@@ -216,7 +217,7 @@ func NewRemoteClusterServiceWatcher(
 	})
 
 	stopper := make(chan struct{})
-	return &RemoteClusterServiceWatcher{
+	rcsw := &RemoteClusterServiceWatcher{
 		serviceMirrorNamespace: serviceMirrorNamespace,
 		link:                   link,
 		remoteAPIClient:        remoteAPI,
@@ -235,9 +236,12 @@ func NewRemoteClusterServiceWatcher(
 		liveness:                 liveness,
 		headlessServicesEnabled:  enableHeadlessSvc,
 		namespaceCreationEnabled: enableNamespaceCreation,
-		// always instantiate the gatewayAlive=true to prevent unexpected service fail fast
-		gatewayAlive: true,
-	}, nil
+	}
+
+	// always instantiate the gatewayAlive=true to prevent unexpected service fail fast
+	rcsw.setGatewayAlive(true)
+
+	return rcsw, nil
 }
 
 func (rcsw *RemoteClusterServiceWatcher) mirrorServiceName(remoteName string) string {
@@ -369,6 +373,14 @@ func (rcsw *RemoteClusterServiceWatcher) getMirrorServiceAnnotations(remoteServi
 	annotations[consts.RemoteResourceVersionAnnotation] = remoteService.ResourceVersion // needed to detect real changes
 
 	return annotations
+}
+
+func (rcsw *RemoteClusterServiceWatcher) getGatewayAlive() bool {
+	return rcsw.gatewayAlive.Load()
+}
+
+func (rcsw *RemoteClusterServiceWatcher) setGatewayAlive(alive bool) {
+	rcsw.gatewayAlive.Store(alive)
 }
 
 // Provides annotations for federated service
@@ -1468,25 +1480,29 @@ func (rcsw *RemoteClusterServiceWatcher) Start(ctx context.Context) error {
 	ev := RepairEndpoints{}
 	rcsw.eventsQueue.Add(&ev)
 
-	go func() {
-		ticker := time.NewTicker(rcsw.repairPeriod)
-		for {
-			select {
-			case <-ticker.C:
-				ev := RepairEndpoints{}
-				rcsw.eventsQueue.Add(&ev)
-			case alive := <-rcsw.liveness:
-				rcsw.log.Debugf("gateway liveness change from %t to %t", rcsw.gatewayAlive, alive)
-				rcsw.gatewayAlive = alive
-				ev := RepairEndpoints{}
-				rcsw.eventsQueue.Add(&ev)
-			case <-rcsw.stopper:
-				return
-			}
-		}
-	}()
+	go rcsw.watchGatewayLiveness()
 
 	return nil
+}
+
+func (rcsw *RemoteClusterServiceWatcher) watchGatewayLiveness() {
+	ticker := time.NewTicker(rcsw.repairPeriod)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ticker.C:
+			ev := RepairEndpoints{}
+			rcsw.eventsQueue.Add(&ev)
+		case alive := <-rcsw.liveness:
+			rcsw.log.Debugf("gateway liveness change from %t to %t", rcsw.getGatewayAlive(), alive)
+			rcsw.setGatewayAlive(alive)
+			ev := RepairEndpoints{}
+			rcsw.eventsQueue.Add(&ev)
+		case <-rcsw.stopper:
+			return
+		}
+	}
 }
 
 // Stop stops watching the cluster and cleans up all mirrored resources
@@ -1786,7 +1802,7 @@ func (rcsw *RemoteClusterServiceWatcher) updateMirrorEndpoints(ctx context.Conte
 }
 
 func (rcsw *RemoteClusterServiceWatcher) updateReadiness(endpoints *corev1.Endpoints) {
-	if !rcsw.gatewayAlive {
+	if !rcsw.getGatewayAlive() {
 		rcsw.log.Warnf("gateway for %s/%s does not have ready addresses; setting addresses to not ready", endpoints.Namespace, endpoints.Name)
 		for i := range endpoints.Subsets {
 			endpoints.Subsets[i].NotReadyAddresses = append(endpoints.Subsets[i].NotReadyAddresses, endpoints.Subsets[i].Addresses...)

--- a/multicluster/service-mirror/cluster_watcher_mirroring_test.go
+++ b/multicluster/service-mirror/cluster_watcher_mirroring_test.go
@@ -433,9 +433,9 @@ func TestLocalNamespaceCreatedAfterServiceExport(t *testing.T) {
 		log:                     logging.WithFields(logging.Fields{"cluster": clusterName}),
 		eventsQueue:             q,
 		requeueLimit:            0,
-		gatewayAlive:            true,
 		headlessServicesEnabled: true,
 	}
+	watcher.setGatewayAlive(true)
 
 	q.Add(&RemoteServiceExported{
 		service: remoteService("service-one", "ns1", "111", map[string]string{
@@ -525,8 +525,8 @@ func TestServiceCreatedGatewayAlive(t *testing.T) {
 		log:             logging.WithFields(logging.Fields{"cluster": clusterName}),
 		eventsQueue:     events,
 		requeueLimit:    0,
-		gatewayAlive:    true,
 	}
+	watcher.setGatewayAlive(true)
 
 	events.Add(&RemoteServiceExported{
 		service: remoteService("svc", "ns", "1", map[string]string{
@@ -567,7 +567,7 @@ func TestServiceCreatedGatewayAlive(t *testing.T) {
 
 	// The gateway is now down which triggers repairing Endpoints on the local
 	// cluster.
-	watcher.gatewayAlive = false
+	watcher.setGatewayAlive(false)
 	events.Add(&RepairEndpoints{})
 	for events.Len() > 0 {
 		watcher.processNextEvent(context.Background())
@@ -683,7 +683,6 @@ func TestServiceCreatedGatewayDown(t *testing.T) {
 		log:             logging.WithFields(logging.Fields{"cluster": clusterName}),
 		eventsQueue:     events,
 		requeueLimit:    0,
-		gatewayAlive:    false,
 	}
 
 	events.Add(&RemoteServiceExported{
@@ -725,7 +724,7 @@ func TestServiceCreatedGatewayDown(t *testing.T) {
 
 	// The gateway is now alive which triggers repairing Endpoints on the
 	// local cluster.
-	watcher.gatewayAlive = true
+	watcher.setGatewayAlive(true)
 	events.Add(&RepairEndpoints{})
 	for events.Len() > 0 {
 		watcher.processNextEvent(context.Background())

--- a/multicluster/service-mirror/cluster_watcher_race_test.go
+++ b/multicluster/service-mirror/cluster_watcher_race_test.go
@@ -1,0 +1,48 @@
+package servicemirror
+
+import (
+	"testing"
+	"time"
+
+	logging "github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/client-go/util/workqueue"
+)
+
+func TestGatewayAliveSynchronization(t *testing.T) {
+	stopper := make(chan struct{})
+	queue := workqueue.NewTypedRateLimitingQueue(workqueue.DefaultTypedControllerRateLimiter[any]())
+	defer queue.ShutDown()
+
+	watcher := &RemoteClusterServiceWatcher{
+		stopper:      stopper,
+		log:          logging.NewEntry(logging.New()),
+		eventsQueue:  queue,
+		repairPeriod: time.Hour,
+		liveness:     make(chan bool, 1024),
+	}
+	watcher.setGatewayAlive(true)
+
+	done := make(chan struct{})
+	go func() {
+		defer close(done)
+		watcher.watchGatewayLiveness()
+	}()
+
+	endpoints := &corev1.Endpoints{
+		Subsets: []corev1.EndpointSubset{{
+			Addresses: []corev1.EndpointAddress{{
+				IP: "192.0.2.1",
+			}},
+		}},
+	}
+
+	for i := 0; i < 1024; i++ {
+		watcher.liveness <- i%2 == 0
+		ep := endpoints.DeepCopy()
+		watcher.updateReadiness(ep)
+	}
+
+	close(stopper)
+	<-done
+}

--- a/multicluster/service-mirror/cluster_watcher_test_util.go
+++ b/multicluster/service-mirror/cluster_watcher_test_util.go
@@ -65,9 +65,9 @@ func (te *testEnvironment) runEnvironment(watcherQueue workqueue.TypedRateLimiti
 		log:                     logging.WithFields(logging.Fields{"cluster": clusterName}),
 		eventsQueue:             watcherQueue,
 		requeueLimit:            0,
-		gatewayAlive:            true,
 		headlessServicesEnabled: true,
 	}
+	watcher.setGatewayAlive(true)
 
 	for _, ev := range te.events {
 		watcherQueue.Add(ev)

